docker/configure_workers_and_start.py

```diff
@@ -364,137 +384,86 @@ def convert(src: str, dst: str, **template_vars: object) -> None:
-                instance_map[worker_name] = {
-                    "path": f"/run/worker.{worker_port}",
-                }
-            else:
-                instance_map[worker_name] = {
-                    "host": "localhost",
-                    "port": worker_port,
-                }
+    if os.environ.get("SYNAPSE_USE_UNIX_SOCKET", False):
+        instance_map[worker_name] = {
+            "path": f"/run/worker.{worker_port}",
+        }
+    else:
+        instance_map[worker_name] = {
+            "host": "localhost",
+            "port": worker_port,
+        }
 
 
 def merge_worker_template_configs(
-    existing_dict: Optional[Dict[str, Any]],
-    to_be_merged_dict: Dict[str, Any],
-) -> Dict[str, Any]:
+    existing_template: WorkerTemplate,
+    to_be_merged_template: WorkerTemplate,
+) -> WorkerTemplate:
     """When given an existing dict of worker template configuration consisting with both
         dicts and lists, merge new template data from WORKERS_CONFIG(or create) and
         return new dict.
 
     Args:
-        existing_dict: Either an existing worker template or a fresh blank one.
-        to_be_merged_dict: The template from WORKERS_CONFIGS to be merged into
+        existing_template: Either an existing worker template or a fresh blank one.
+        to_be_merged_template: The template from WORKERS_CONFIGS to be merged into
             existing_dict.
     Returns: The newly merged together dict values.
     """
-    new_dict: Dict[str, Any] = {}
-    if not existing_dict:
-        # It doesn't exist yet, just use the new dict(but take a copy not a reference)
-        new_dict = to_be_merged_dict.copy()
-    else:
-        for i in to_be_merged_dict.keys():
-            if (i == "endpoint_patterns") or (i == "listener_resources"):
-                # merge the two lists, remove duplicates
-                new_dict[i] = list(set(existing_dict[i] + to_be_merged_dict[i]))
-            elif i == "shared_extra_conf":
-                # merge dictionary's, the worker name will be replaced later
-                new_dict[i] = {**existing_dict[i], **to_be_merged_dict[i]}
-            elif i == "worker_extra_conf":
-                # There is only one worker type that has a 'worker_extra_conf' and it is
-                # the media_repo. Since duplicate worker types on the same worker don't
-                # work, this is fine.
-                new_dict[i] = existing_dict[i] + to_be_merged_dict[i]
-            else:
-                # Everything else should be identical, like "app", which only works
-                # because all apps are now generic_workers.
-                new_dict[i] = to_be_merged_dict[i]
-    return new_dict
+    # copy existing_template without any replacements
+    new_template: WorkerTemplate = dataclasses.replace(existing_template)
+
+    # add listener resources from the other set
+    new_template.listener_resources |= to_be_merged_template.listener_resources
```

Comment:
Will this potentially modify `existing_template.listener_resources`?

Reply:
I am disturbed. I thought `... |= ...` was defined to be the same as `... = ... | ...`. You rightly point out that it is not (!!!):

```python
>>> a = {1}
>>> b = {2}
>>> c = a
>>> a | b
{1, 2}
>>> c
{1}
>>> a |= b
>>> c
{1, 2}

# ^ !!!

>>> a = {1}
>>> c = a
>>> a = a | b
>>> c
{1}
>>> a
{1, 2}
```

Reply:
Yeah, all those sort of operators update in place, rather than creating a new object with the same name. It's true that `a |= b` means that `a` will now contain the same thing as `a | b`, but that is obviously different to `a = a | b` (which is creating a new object and assigning it the name `a`)

Reply:
'obvious' now I've been burnt I guess but I think I probably learnt about this operator from some other language where it genuinely was syntactic sugar, but maybe only applied to simple types where that wouldn't matter anyway.